ext/standaloneusers/Civi/Api4/Action/User/Save.php

```diff
@@ -2,8 +2,37 @@
 namespace Civi\Api4\Action\User;
 
 use Civi\Api4\Generic\DAOSaveAction;
+use Civi\Api4\Generic\Result;
+use Civi\API\Exception\UnauthorizedException;
+use CRM_Core_Permission;
 
 class Save extends DAOSaveAction {
   use WriteTrait;
 
+  /**
+   * Run some permission checks before doing the save.
+   */
+  public function _run(Result $result) {
+    if (empty($this->records)) {
+      return;
+    }
+    if ($this->getCheckPermissions()) {
+      $loggedInUfID = (int) \CRM_Utils_System::getLoggedInUfID();
+      if (!$loggedInUfID) {
+        // Never allow save if not logged in.
+        throw new UnauthorizedException("User.save API call when not logged in.");
```

Comment:
I see what this is trying to do, but I think there's probably a cleaner way to do it.
Overriding the `_run` function in APIv4 is discouraged because it's such a kludge & there are finer-grained ways to insert specific logic. In this case you're doing permission checks. That should happen here in the validateValues function:

https://github.com/civicrm/civicrm-core/blob/283139dc0dab8556bd644357ef7b9cd8fff34f7e/Civi/Api4/Generic/AbstractSaveAction.php#L75-L79

But rather than override that function, you should be able to respond to one of the two event it emits, either this one:
https://github.com/civicrm/civicrm-core/blob/283139dc0dab8556bd644357ef7b9cd8fff34f7e/Civi/Api4/Generic/AbstractSaveAction.php#L115
or this one:
https://github.com/civicrm/civicrm-core/blob/d5a76d577c2d11fd0e524424159d0192a68a6790/Civi/Api4/Utils/CoreUtil.php#L253

Actually, it looks like you already are responding to the second one in the BAO... so I don't think this override is necessary at all!